sematic/external_resource.py

```diff
@@ -0,0 +1,224 @@
+        The resource is being deactivated, and likely is no longer usable.
+    INACTIVE:
+        The resource has been deactivated and may not even exist anymore.
+        It is not usable.
+    """
+
+    CREATED = "CREATED"
+    ACTIVATING = "ACTIVATING"
+    ACTIVE = "ACTIVE"
+    DEACTIVATING = "DEACTIVATING"
+    INACTIVE = "INACTIVE"
+
+    def is_allowed_transition(self, other_state: "ResourceStatus") -> bool:
+        """True if going from the current state to the other is allowed, otherwise False.
+
+        Parameters
+        ----------
+        other_state:
+            The state being transitioned to
+
+        Returns
+        -------
+        True if going from the current state to the other is allowed, otherwise False.
+        """
+        return other_state in _ALLOWED_TRANSITIONS[self]
+
+    def is_terminal(self) -> bool:
+        """True if there are no states that can follow this one, False otherwise."""
+        return len(_ALLOWED_TRANSITIONS[self]) == 0
+
+
+_ALLOWED_TRANSITIONS = {
+    None: {ResourceState.CREATED},
+    ResourceState.CREATED: {
+        # Created -> Activating: normal progression for activation
+        ResourceState.ACTIVATING,
+    },
+    ResourceState.ACTIVATING: {
+        # Activating -> Active: normal progression for successful activation
+        ResourceState.ACTIVE,
+        # Activating -> Deactivating: activation failed, immediate deactivation
+        ResourceState.DEACTIVATING,
+    },
+    ResourceState.ACTIVE: {
+        # Active -> Deactivating:
+        #   - possibly normal termination, due to no longer being needed.
+        #   - possibly an error with the resource. Status message should have more info.
+        ResourceState.DEACTIVATING,
+    },
+    ResourceState.DEACTIVATING: {
+        # Deactivating -> Inactive: normal progression for deactivation
+        ResourceState.INACTIVE,
+    },
+    ResourceState.DEACTIVATING: {},
+}
+
+
+@dataclass(frozen=True)
+class ResourceStatus:
+    """The status of the resource
+
+    Attributes
+    ----------
+    state:
+        The discrete state that the resource is in
+    message:
+        A human-readable message about why the resource entered this state
+    last_update_epoch_time:
+        The last time this status was checked against the actual external resource
+    """
+
+    state: ResourceState
+    message: str
+    last_update_epoch_time: int
+
+
+@dataclass(frozen=True)
+class ExternalResource(AbstractExternalResource):
+    """Represents a resource tracked by Sematic for usage in Sematic funcs.
+
+    Examples of possible external resources include small data processing
+    clusters or distributed training clusters.
+
+    Specific external resources should inherit from this class. They may add their
+    own dataclass attributes and must implement the _do_activate, _do_deactivate, and
+    _do_update methods on this class.
+
+    This class represents immutable objects. As with immutable strings, the
```

Comment:
I do a lot of validation. These are the kinds of things where it's nice to be able to rely on invariants that you have enforced as you reason about what's happened in the system. "What happened to that Spark cluster?" is something we should ideally always be able to answer fairly easily, at least as far as it concerns our own handling of said cluster.